### ISSUE TITLE
copy -a: Copy the annotations from each image manifest into the image index [feature request]

### ISSUE THREAD

**Issue by sjackman** [containers/skopeo#1232]
When using `skopeo copy -a` I would like the annotations from each `application/vnd.oci.image.manifest.v1+json` copied into the `application/vnd.oci.image.index.v1+json`, in particular the `org.opencontainers.image.ref.name`. See the following example: https://github.com/opencontainers/image-spec/blob/master/image-layout.md#index-example

**Comment by vrothberg** [comment]
Thanks for reaching out!

To elaborate on the request:
* All kinds of descriptors support annotations.
* An image index (`application/vnd.oci.image.index.v1+json`) allows for specifying annotations for each "pointer" to a manifest.
* A manifest (`application/vnd.oci.image.manifest.v1+json`) itself can also specify annotations. Those are semantically different from the ones the top-level image index may ship with.

Currently, when running `skopeo -a` the image index and the manifests are copied verbatim.  This requests wants to further manipulate the image index and "copy up" the annotations of each manifest into the specific entry in the image index.

Is that what you're looking for?

**Comment by sjackman** [comment]
```console
$ skopeo --version
skopeo version 1.2.2
$ skopeo copy -a oci:ssm--1.4 oci:ssm--1.4-2
Getting image list signatures
Copying 2 of 2 images in list
Copying image sha256:771515de308d684e2f7a92be3d592010dfd410941da63dcf0382a2662c5b467b (1/2)
Getting image source signatures
Copying blob 1b88c88c5b51 done  
Copying config 38e6e8b622 done  
Writing manifest to image destination
Storing signatures
Copying image sha256:54b93c74a268430fbb0c7c87fa981ba90f15934732bdcabd57ed184bbfe3b8aa (2/2)
Getting image source signatures
Copying blob 0cea947187c8 done  
Copying config a0cb902d0c done  
Writing manifest to image destination
Storing signatures
Writing manifest list to image destination
Storing list signatures
$ jq '[.manifests[].digest,.manifests[].size,input_filename]' ssm--1.4/index.json ssm--1.4-2/index.json
```
```json
[
  "sha256:614dcc39ccf41223c53d900c415a0cfd39fe8f376668722dcbf23f14e99d70a9",
  1424,
  "ssm--1.4/index.json"
]
[
  "sha256:a727153188bb2d25d6f03a897e61de2971d4dc996cf5581c9ee7179cf8065ad6",
  1120,
  "ssm--1.4-2/index.json"
]
```

> Currently, when running `skopeo -a` the image index and the manifests are copied verbatim.

I'd be fine with the `application/vnd.oci.image.index.v1+json` being copied verbatim, but what I'm seeing is it being rewritten.

> This requests wants to further manipulate the image index and "copy up" the annotations of each manifest into the specific entry in the image index. Is that what you're looking for?

Yes, that is the feature request. Copying the image index verbatim would be also be fine.

Thank you very much for Skopeo! It's exactly the tool that I needed.

### `ssm--1.4/blobs/sha256/614dcc39ccf41223c53d900c415a0cfd39fe8f376668722dcbf23f14e99d70a9`
```json
{
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "schemaVersion": 2,
  "manifests": [
    {
      "digest": "sha256:771515de308d684e2f7a92be3d592010dfd410941da63dcf0382a2662c5b467b",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": {
        "architecture": "amd64",
        "os": "linux",
        "os.version": "2.23"
      },
      "size": 1171
    },
    {
      "digest": "sha256:54b93c74a268430fbb0c7c87fa981ba90f15934732bdcabd57ed184bbfe3b8aa",
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "platform": {
        "architecture": "amd64",
        "os": "darwin",
        "os.version": "10.15"
      },
      "size": 1167
    }
  ],
  "annotations": {
    "org.opencontainers.image.created": "2021-02-19",
    "org.opencontainers.image.description": "Secondary-structure matching, tool for fast protein structure alignment",
    "org.opencontainers.image.license": "GPL-3.0-only",
    "org.opencontainers.image.revision": "f4b62b94e7a53aef8ec832fc9e1f1c1c00247243fa3d302ebbd21d20ccf75a17b8a9918d2b0801a8",
    "org.opencontainers.image.source": "https://github.com/brewsci/homebrew-bio/blob/f4b62b94e7a53aef8ec832fc9e1f1c1c00247243/Formula/ssm.rb",
    "org.opencontainers.image.title": "brewsci/bio/ssm",
    "org.opencontainers.image.url": "https://www2.mrc-lmb.cam.ac.uk",
    "org.opencontainers.image.version": "1.4"
  }
}
```

### `ssm--1.4-2/blobs/sha256/a727153188bb2d25d6f03a897e61de2971d4dc996cf5581c9ee7179cf8065ad6`
```json
{"schemaVersion":2,"manifests":[{"mediaType":"application/vnd.oci.image.manifest.v1+json","digest":"sha256:771515de308d684e2f7a92be3d592010dfd410941da63dcf0382a2662c5b467b","size":1025,"platform":{"architecture":"amd64","os":"linux","os.version":"2.23"}},{"mediaType":"application/vnd.oci.image.manifest.v1+json","digest":"sha256:54b93c74a268430fbb0c7c87fa981ba90f15934732bdcabd57ed184bbfe3b8aa","size":1021,"platform":{"architecture":"amd64","os":"darwin","os.version":"10.15"}}],"annotations":{"org.opencontainers.image.created":"2021-02-19","org.opencontainers.image.description":"Secondary-structure matching, tool for fast protein structure alignment","org.opencontainers.image.license":"GPL-3.0-only","org.opencontainers.image.revision":"f4b62b94e7a53aef8ec832fc9e1f1c1c00247243fa3d302ebbd21d20ccf75a17b8a9918d2b0801a8","org.opencontainers.image.source":"https://github.com/brewsci/homebrew-bio/blob/f4b62b94e7a53aef8ec832fc9e1f1c1c00247243/Formula/ssm.rb","org.opencontainers.image.title":"brewsci/bio/ssm","org.opencontainers.image.url":"https://www2.mrc-lmb.cam.ac.uk","org.opencontainers.image.version":"1.4"}}
```

**Comment by mtrmac** [comment]
(Drive-by comment:) It’s unclear to me what is the request here.

Looking at the 614dcc39ccf41223c53d900c415a0cfd39fe8f376668722dcbf23f14e99d70a9 and a727153188bb2d25d6f03a897e61de2971d4dc996cf5581c9ee7179cf8065ad6 digests, that seems to be reporting the outcome of that the `skopeo copy -a` command above. But both of the OCI indices have _exactly the same_ annotations. Isn’t that what you were asking for?

A difference is that the copy has probably (we don’t have the full data) converted the member manifest from v2s2 to OCI. That doesn’t look immediately surprising, v2s2 is not valid OCI.

---

(As for copy-up, `skopeo copy`’s job is not primarily to restructure/edit images. That looks like something more appropriate for Buildah and its code to create multi-arch indices from the component images. OTOH `copy` does do format conversions and the like, so it’s not _completely_ inconceivable but conceptually it should fit in the idea of transferring an image as a single unit.)

---

Just _very vaguely_ guessing, is your complaint that the original image manifests have annotations, and those have been discarded during the copy as part of the conversion to OCI manifests? Note that the original v2s2 format does not define annotations: https://github.com/distribution/distribution/blob/main/docs/spec/manifest-v2-2.md .

Or are you using the OCI image index as a way to store multiple _completely unrelated_ images, and you want `skopeo copy -a` to copy the whole set? That’s not what that command really does, that command exists to copy a _single multi-arch image_ (which is also expressed as an OCI image index).

**Comment by sjackman** [comment]
I wrote up this issue with a way under-described feature request, and then when I replied, I had confused this issue with a totally different one, so my reply made no sense at all. 😬 Really sorry for the noise! What I wanted was to copy the annotations from the source OCI image manifests to the resulting OCI image index in `manifests[].annotations`, as in this https://github.com/opencontainers/image-spec/blob/master/image-layout.md#index-example. I can think of other ways to copy that data though, so feel free to close this issue.
